api/src/views/mod.rs

```diff
@@ -22,5 +22,12 @@ pub use user_program::*;
 pub trait View {
     type Output;
 
-    fn execute(&self) -> ResponseResult<Self::Output>;
+    fn check_permissions(&self) -> ResponseResult<()>;
+
+    fn execute_internal(&self) -> ResponseResult<Self::Output>;
+
+    fn execute(&self) -> ResponseResult<Self::Output> {
+        self.check_permissions()?;
+        self.execute_internal()
+    }
```

Comment:
Is it possible to have two associated types? It seems `check_permissions` always throws the same error so maybe we could set the second associated type to the desired permissions so we don't need to do the boiler plate in each implementation

Reply:
it would be weird for the few impls that always return Ok(), so maybe we could just define a lambda that does the perm check? Might be overkill though so im cool with not doing any of this

Reply:
Yes, multiple associated types is possible, but I have a bigger plan for this. In general we'll have two different ways of rejecting based on permissions. One is "you're allowed to know this thing exists, but you can't do whatever you tried", e.g. if you try to modify a hardware spec since those are public but read only except for us.

The other option is that you shouldn't even know something exists, e.g. if you try to delete someone else's user program (since you do that by UUID). In those cases, we want to pretend like the object doesn't even exist, to prevent data leaks.

My long term plan is to convert this views into sort of "view sets" (that's what Django Rest Framework calls them). You'd have one view set per "resource" (basically per DB table), and each view set would define functionality for reading 1 object, reading multiple, and whatever other mutations are possible. So for example, you'd have one view set for user programs, and as part of that you'd define a function that checks if a user is allowed to read a particular user program. So when you go to mutate a user program, you check for read perm first, and if they don't have that, you pretend like it doesn't exist. But if they have read, then you move onto the write perm check, and if that fails you return `PermissionDenied`.

Does that make sense? I haven't fully fleshed out the idea but that's the gist.

Reply:
And with that new scheme it should reduce a lot of the boilerplate

Reply:
that mostly makes sense, tldr would be this is just be the first check before doing more work?